library/alloc/src/vec.rs

```diff
@@ -1410,6 +1410,18 @@ impl<T> Vec<T> {
             assert_failed(at, self.len());
         }
 
+        if at == 0 {
+            // the new vector can take over the original buffer and avoid the copy
+            let buf = &self.buf;
+            let capacity = buf.capacity();
+            let other = unsafe {
+                Vec { buf: RawVec::from_raw_parts(buf.ptr(), buf.capacity()), len: self.len }
+            };
+            mem::forget(mem::replace(&mut self.buf, RawVec::with_capacity(capacity)));
+            self.len = 0;
+            return other;
```

Comment:
Is this significantly different from `return mm::replace(&mut self, Vec::with_capacity(self.capacity())`?

Reply:
ooh, that might work. I'll try it.

Reply:
Done